matchzoo/preprocessors.py

```diff
@@ -1,146 +1,165 @@
-    """
-    return [token for token in tokens if not match_punc.search(token)]
 
+class DigitRemovalUnit(StatelessProcessUnit):
+    """Process unit to remove digits."""
 
-def remove_digits(tokens: list) -> list:
-    """
-    Remove digits from list of tokens.
+    def transform(self, tokens: list) -> list:
+        """
+        Remove digits from list of tokens.
 
-    :param tokens: list of tokens to be filtered.
+        :param tokens: list of tokens to be filtered.
 
-    :return tokens: tokens of tokens without digits.
-    """
-    return [token for token in tokens if not token.isdigit()]
+        :return tokens: tokens of tokens without digits.
+        """
+        return [token for token in tokens if not token.isdigit()]
 
 
-def stemming(tokens: list, stemmer: str = 'porter') -> list:
-    """
-    Reducing inflected words to their word stem, base or root form.
+class StopRemovalUnit(StatelessProcessUnit):
+    """Process unit to remove stop words."""
 
-    :param tokens: list of string to be stemmed.
+    def transform(self, tokens: list, **kwargs: dict) -> list:
+        """
+        Remove stopwords from list of tokenized tokens.
 
-    :raise ValueError: stemmer type should be porter or lancaster.
+        :param tokens: list of tokenized tokens.
+        :param lang: language code for stopwords.
 
-    :return tokens: stemmed token.
-    """
-    if stemmer == 'porter':
-        porter_stemmer = nltk.stem.PorterStemmer()
-        return [porter_stemmer.stem(token) for token in tokens]
-    elif stemmer == 'lancaster' or stemmer == 'krovetz':
-        lancaster_stemmer = nltk.stem.lancaster.LancasterStemmer()
-        return [lancaster_stemmer.stem(token) for token in tokens]
-    else:
-        raise ValueError(
-            'Not supported supported stemmer type: {}'.format(stemmer))
+        :return tokens: list of tokenized tokens without stopwords.
+        """
+        lang = kwargs.get('lang', 'en')
+        return [token
+                for token
+                in tokens
+                if token not in StopRemovalUnit._get_stopwords(lang)]
 
+    @staticmethod
+    def _get_stopwords(lang: str) -> list:
+        """
+        Get stopwords based on language.
 
-def lemmatization(tokens: list) -> list:
-    """
-    Lemmatization a sequence of tokens.
+        :params lang: language code.
 
-    :param tokens: list of tokens to be lemmatized.
+        :return stop_list: list of stop words.
+        """
+        return many_stop_words.get_stop_words(lang)
 
-    :return tokens: list of lemmatizd tokens.
-    """
-    lemmatizer = nltk.WordNetLemmatizer()
-    return [lemmatizer.lemmatize(token, pos='v') for token in tokens]
 
+class StemmingUnit(StatelessProcessUnit):
+    """Process unit for token stemming."""
 
-def chain(*funcs: typing.Callable) -> typing.Callable:
-    """
-    Chain a list of functions execute as pipeline.
+    def transform(self, tokens: list, **kwargs: dict) -> list:
```

Comment:
Same, move the configs into `__init__`. Think about instantiate one instance, and apply on both training data and testing data. It removes repetition in this very common usage.

Reply:
okay